src/python/pants/engine/rules.py

```diff
@@ -230,6 +156,9 @@ def rule_decorator(func, **kwargs) -> Callable:
     if not inspect.isfunction(func):
         raise ValueError("The @rule decorator expects to be placed on a function.")
 
+    if hasattr(func, "rule_helper"):
+        raise ValueError("A @rule cannot be a @rule_helper")
```

Comment:
Technically I think this would be more correct as `"A @rule_helper cannot be a @rule"`, as the rule_helper decorator has already been applied.